File: concepts/data_structures/go/data_structures.go
// データ構造 - Go 実装
//
// Go の組み込みデータ構造と自作実装

package main

import (
	"container/heap"
	"container/list"
	"fmt"
	"sync"
)

func main() {
	fmt.Println("=== Data Structures Demo ===\n")

	slices()
	maps()
	linkedList()
	stackAndQueue()
	priorityQueue()
	set()
	syncMap()
}

// --- スライス ---
func slices() {
	fmt.Println("--- スライス ---")

	// 作成
	s1 := []int{1, 2, 3, 4, 5}
	s2 := make([]int, 5)    // len=5, cap=5
	s3 := make([]int, 0, 5) // len=0, cap=5

	fmt.Printf("s1: %v, len=%d, cap=%d\n", s1, len(s1), cap(s1))
	fmt.Printf("s2: %v, len=%d, cap=%d\n", s2, len(s2), cap(s2))
	fmt.Printf("s3: %v, len=%d, cap=%d\n", s3, len(s3), cap(s3))

	// 追加
	s3 = append(s3, 1, 2, 3)
	fmt.Printf("After append: %v\n", s3)

	// スライシング
	fmt.Printf("s1[1:3]: %v\n", s1[1:3])
	fmt.Printf("s1[:3]:  %v\n", s1[:3])
	fmt.Printf("s1[2:]:  %v\n", s1[2:])

	// コピー
	dst := make([]int, len(s1))
	copy(dst, s1)
	dst[0] = 100
	fmt.Printf("s1: %v, dst: %v\n", s1, dst)

	// 削除 (index 2)
	idx := 2
	s1 = append(s1[:idx], s1[idx+1:]...)
	fmt.Printf("After delete index 2: %v\n", s1)

	fmt.Println()
}

// --- マップ ---
func maps() {
	fmt.Println("--- マップ ---")

	// 作成
	m := map[string]int{
		"one":   1,
		"two":   2,
		"three": 3,
	}
	fmt.Printf("map: %v\n", m)

	// アクセス
	fmt.Printf("m[\"one\"]: %d\n", m["one"])

	// 存在確認
	val, ok := m["four"]
	fmt.Printf("m[\"four\"]: %d, exists: %t\n", val, ok)

	// 追加/更新
	m["four"] = 4
	fmt.Printf("After add: %v\n", m)

	// 削除
	delete(m, "one")
	fmt.Printf("After delete: %v\n", m)

	// イテレーション
	fmt.Print("Keys: ")
	for k := range m {
		fmt.Printf("%s ", k)
	}
	fmt.Println()

	fmt.Println()
}

// --- 連結リスト (container/list) ---
func linkedList() {
	fmt.Println("--- 連結リスト ---")

	// 標準ライブラリの双方向リスト
	l := list.New()

	// 追加
	l.PushBack(1)
	l.PushBack(2)
	l.PushBack(3)
	l.PushFront(0)

	// イテレーション
	fmt.Print("List: ")
	for e := l.Front(); e != nil; e = e.Next() {
		fmt.Printf("%v ", e.Value)
	}
	fmt.Println()

	// 削除
	front := l.Front()
	l.Remove(front)
	fmt.Print("After remove front: ")
	for e := l.Front(); e != nil; e = e.Next() {
		fmt.Printf("%v ", e.Value)
	}
	fmt.Println()

	// 自作の単方向リスト
	fmt.Println("\nCustom singly linked list:")
	myList := &SinglyLinkedList[int]{}
	myList.PushFront(3)
	myList.PushFront(2)
	myList.PushFront(1)
	myList.PushBack(4)
	fmt.Printf("List: %v\n", myList.ToSlice())
	fmt.Printf("PopFront: %v\n", myList.PopFront())
	fmt.Printf("After pop: %v\n", myList.ToSlice())

	fmt.Println()
}

// 自作単方向リスト
type Node[T any] struct {
	Value T
	Next  *Node[T]
}

type SinglyLinkedList[T any] struct {
	Head *Node[T]
	size int
}

func (l *SinglyLinkedList[T]) PushFront(value T) {
	l.Head = &Node[T]{Value: value, Next: l.Head}
	l.size++
}

func (l *SinglyLinkedList[T]) PushBack(value T) {
	node := &Node[T]{Value: value}
	if l.Head == nil {
		l.Head = node
	} else {
		current := l.Head
		for current.Next != nil {
			current = current.Next
		}
		current.Next = node
	}
	l.size++
}

func (l *SinglyLinkedList[T]) PopFront() T {
	if l.Head == nil {
		var zero T
		return zero
	}
	value := l.Head.Value
	l.Head = l.Head.Next
	l.size--
	return value
}

func (l *SinglyLinkedList[T]) ToSlice() []T {
	result := make([]T, 0, l.size)
	for current := l.Head; current != nil; current = current.Next {
		result = append(result, current.Value)
	}
	return result
}

// --- スタックとキュー ---
func stackAndQueue() {
	fmt.Println("--- スタックとキュー ---")

	// スタック (スライスで実装)
	fmt.Println("Stack:")
	stack := &Stack[int]{}
	stack.Push(1)
	stack.Push(2)
	stack.Push(3)
	fmt.Printf("  Stack: %v\n", stack.items)
	fmt.Printf("  Pop: %v\n", stack.Pop())
	fmt.Printf("  Peek: %v\n", stack.Peek())

	// キュー (スライスで実装)
	fmt.Println("\nQueue:")
	queue := &Queue[string]{}
	queue.Enqueue("first")
	queue.Enqueue("second")
	queue.Enqueue("third")
	fmt.Printf("  Queue: %v\n", queue.items)
	fmt.Printf("  Dequeue: %v\n", queue.Dequeue())
	fmt.Printf("  Front: %v\n", queue.Front())

	fmt.Println()
}

// スタック
type Stack[T any] struct {
	items []T
}

func (s *Stack[T]) Push(item T) {
	s.items = append(s.items, item)
}

func (s *Stack[T]) Pop() T {
	if len(s.items) == 0 {
		var zero T
		return zero
	}
	item := s.items[len(s.items)-1]
	s.items = s.items[:len(s.items)-1]
	return item
}

func (s *Stack[T]) Peek() T {
	if len(s.items) == 0 {
		var zero T
		return zero
	}
	return s.items[len(s.items)-1]
}

func (s *Stack[T]) IsEmpty() bool {
	return len(s.items) == 0
}

// キュー
type Queue[T any] struct {
	items []T
}

func (q *Queue[T]) Enqueue(item T) {
	q.items = append(q.items, item)
}

func (q *Queue[T]) Dequeue() T {
	if len(q.items) == 0 {
		var zero T
		return zero
	}
	item := q.items[0]
	q.items = q.items[1:]
	return item
}

func (q *Queue[T]) Front() T {
	if len(q.items) == 0 {
		var zero T
		return zero
	}
	return q.items[0]
}

func (q *Queue[T]) IsEmpty() bool {
	return len(q.items) == 0
}

// --- 優先度キュー (container/heap) ---
func priorityQueue() {
	fmt.Println("--- 優先度キュー ---")

	// container/heap を使った優先度キュー
	pq := &IntHeap{3, 1, 4, 1, 5, 9, 2, 6}
	heap.Init(pq)

	heap.Push(pq, 0)

	fmt.Print("Pop order (min heap): ")
	for pq.Len() > 0 {
		fmt.Printf("%d ", heap.Pop(pq))
	}
	fmt.Println()

	fmt.Println()
}

// heap.Interface を実装
type IntHeap []int

func (h IntHeap) Len() int           { return len(h) }
func (h IntHeap) Less(i, j int) bool { return h[i] < h[j] }
func (h IntHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }

func (h *IntHeap) Push(x any) {
	*h = append(*h, x.(int))
}

func (h *IntHeap) Pop() any {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}

// --- Set ---
func set() {
	fmt.Println("--- Set (map[T]struct{}) ---")

	// Go には組み込みの Set がないので、map で代用
	set1 := make(Set[int])
	set1.Add(1)
	set1.Add(2)
	set1.Add(3)

	set2 := make(Set[int])
	set2.Add(2)
	set2.Add(3)
	set2.Add(4)

	fmt.Printf("set1: %v\n", set1.ToSlice())
	fmt.Printf("set2: %v\n", set2.ToSlice())
	fmt.Printf("set1.Contains(2): %t\n", set1.Contains(2))
	fmt.Printf("Union: %v\n", set1.Union(set2).ToSlice())
	fmt.Printf("Intersection: %v\n", set1.Intersection(set2).ToSlice())

	fmt.Println()
}

type Set[T comparable] map[T]struct{}

func (s Set[T]) Add(item T) {
	s[item] = struct{}{}
}

func (s Set[T]) Remove(item T) {
	delete(s, item)
}

func (s Set[T]) Contains(item T) bool {
	_, ok := s[item]
	return ok
}

func (s Set[T]) ToSlice() []T {
	result := make([]T, 0, len(s))
	for k := range s {
		result = append(result, k)
	}
	return result
}

func (s Set[T]) Union(other Set[T]) Set[T] {
	result := make(Set[T])
	for k := range s {
		result.Add(k)
	}
	for k := range other {
		result.Add(k)
	}
	return result
}

func (s Set[T]) Intersection(other Set[T]) Set[T] {
	result := make(Set[T])
	for k := range s {
		if other.Contains(k) {
			result.Add(k)
		}
	}
	return result
}

// --- sync.Map ---
func syncMap() {
	fmt.Println("--- sync.Map (並行安全) ---")

	var m sync.Map

	// 書き込み
	m.Store("key1", "value1")
	m.Store("key2", "value2")

	// 読み取り
	if val, ok := m.Load("key1"); ok {
		fmt.Printf("key1: %v\n", val)
	}

	// 存在しなければ保存
	actual, loaded := m.LoadOrStore("key3", "value3")
	fmt.Printf("key3: %v, loaded: %t\n", actual, loaded)

	// イテレーション
	fmt.Print("All keys: ")
	m.Range(func(key, value any) bool {
		fmt.Printf("%v ", key)
		return true
	})
	fmt.Println()
}

File: concepts/data_structures/cpp/data_structures.cpp
// データ構造 - C++ 実装
//
// C++ STL には豊富なコンテナが用意されている。
// vector, list, deque, map, set, unordered_map など。

#include <iostream>
#include <vector>
#include <array>
#include <list>
#include <forward_list>
#include <deque>
#include <stack>
#include <queue>
#include <map>
#include <unordered_map>
#include <set>
#include <unordered_set>
#include <string>
#include <algorithm>
#include <memory>

using namespace std;

// === vector ===
void vector_demo() {
    cout << "=== vector ===" << endl;

    // 作成
    vector<int> v = {1, 2, 3, 4, 5};
    cout << "  vector: ";
    for (int x : v) cout << x << " ";
    cout << endl;

    // 操作
    v.push_back(6);
    v.insert(v.begin(), 0);
    cout << "  after push_back/insert: ";
    for (int x : v) cout << x << " ";
    cout << endl;

    // アクセス
    cout << "  v[2] = " << v[2] << endl;
    cout << "  v.at(2) = " << v.at(2) << endl;
    cout << "  front = " << v.front() << ", back = " << v.back() << endl;

    // サイズ
    cout << "  size = " << v.size() << ", capacity = " << v.capacity() << endl;

    // 削除
    v.pop_back();
    v.erase(v.begin());
    cout << "  after pop_back/erase: ";
    for (int x : v) cout << x << " ";
    cout << endl;

    cout << endl;
}

// === array ===
void array_demo() {
    cout << "=== array ===" << endl;

    // 固定長配列
    array<int, 5> arr = {1, 2, 3, 4, 5};

    cout << "  array: ";
    for (int x : arr) cout << x << " ";
    cout << endl;

    cout << "  size = " << arr.size() << endl;
    cout << "  arr[2] = " << arr[2] << endl;

    // fill
    arr.fill(0);
    cout << "  after fill(0): ";
    for (int x : arr) cout << x << " ";
    cout << endl;

    cout << endl;
}

// === list (双方向リスト) ===
void list_demo() {
    cout << "=== list (双方向リスト) ===" << endl;

    list<int> lst = {1, 2, 3};

    lst.push_front(0);
    lst.push_back(4);
    cout << "  list: ";
    for (int x : lst) cout << x << " ";
    cout << endl;

    lst.pop_front();
    lst.pop_back();
    cout << "  after pop_front/pop_back: ";
    for (int x : lst) cout << x << " ";
    cout << endl;

    // splice
    list<int> other = {10, 20};
    lst.splice(lst.end(), other);
    cout << "  after splice: ";
    for (int x : lst) cout << x << " ";
    cout << endl;

    cout << endl;
}

// === deque ===
void deque_demo() {
    cout << "=== deque ===" << endl;

    deque<int> d = {2, 3, 4};

    d.push_front(1);
    d.push_back(5);
    cout << "  deque: ";
    for (int x : d) cout << x << " ";
    cout << endl;

    cout << "  front = " << d.front() << ", back = " << d.back() << endl;

    d.pop_front();
    d.pop_back();
    cout << "  after pops: ";
    for (int x : d) cout << x << " ";
    cout << endl;

    cout << endl;
}

// === stack ===
void stack_demo() {
    cout << "=== stack ===" << endl;

    stack<int> s;
    s.push(1);
    s.push(2);
    s.push(3);

    cout << "  top = " << s.top() << endl;
    cout << "  size = " << s.size() << endl;

    cout << "  popping: ";
    while (!s.empty()) {
        cout << s.top() << " ";
        s.pop();
    }
    cout << endl;

    cout << endl;
}

// === queue ===
void queue_demo() {
    cout << "=== queue ===" << endl;

    queue<int> q;
    q.push(1);
    q.push(2);
    q.push(3);

    cout << "  front = " << q.front() << ", back = " << q.back() << endl;

    cout << "  dequeuing: ";
    while (!q.empty()) {
        cout << q.front() << " ";
        q.pop();
    }
    cout << endl;

    cout << endl;
}

// === priority_queue ===
void priority_queue_demo() {
    cout << "=== priority_queue ===" << endl;

    // max heap (default)
    priority_queue<int> max_pq;
    max_pq.push(3);
    max_pq.push(1);
    max_pq.push(4);
    max_pq.push(1);
    max_pq.push(5);

    cout << "  max heap: ";
    while (!max_pq.empty()) {
        cout << max_pq.top() << " ";
        max_pq.pop();
    }
    cout << endl;

    // min heap
    priority_queue<int, vector<int>, greater<int>> min_pq;
    min_pq.push(3);
    min_pq.push(1);
    min_pq.push(4);
    min_pq.push(1);
    min_pq.push(5);

    cout << "  min heap: ";
    while (!min_pq.empty()) {
        cout << min_pq.top() << " ";
        min_pq.pop();
    }
    cout << endl;

    cout << endl;
}

// === map ===
void map_demo() {
    cout << "=== map ===" << endl;

    map<string, int> m;
    m["one"] = 1;
    m["two"] = 2;
    m["three"] = 3;

    cout << "  map contents:" << endl;
    for (const auto& [key, value] : m) {
        cout << "    " << key << " => " << value << endl;
    }

    // 検索
    cout << "  m[\"two\"] = " << m["two"] << endl;
    cout << "  m.count(\"four\") = " << m.count("four") << endl;

    if (auto it = m.find("three"); it != m.end()) {
        cout << "  found: " << it->first << " => " << it->second << endl;
    }

    // insert_or_assign (C++17)
    m.insert_or_assign("four", 4);

    cout << endl;
}

// === unordered_map ===
void unordered_map_demo() {
    cout << "=== unordered_map ===" << endl;

    unordered_map<string, int> um = {
        {"apple", 100},
        {"banana", 200},
        {"cherry", 300}
    };

    cout << "  unordered_map contents:" << endl;
    for (const auto& [key, value] : um) {
        cout << "    " << key << " => " << value << endl;
    }

    cout << "  um[\"apple\"] = " << um["apple"] << endl;
    cout << "  bucket_count = " << um.bucket_count() << endl;
    cout << "  load_factor = " << um.load_factor() << endl;

    cout << endl;
}

// === set ===
void set_demo() {
    cout << "=== set ===" << endl;

    set<int> s = {3, 1, 4, 1, 5, 9, 2, 6};  // 重複除去、ソート済み

    cout << "  set: ";
    for (int x : s) cout << x << " ";
    cout << endl;

    s.insert(10);
    s.erase(4);

    cout << "  after insert/erase: ";
    for (int x : s) cout << x << " ";
    cout << endl;

    cout << "  contains 5: " << boolalpha << s.contains(5) << endl;

    // lower_bound, upper_bound
    cout << "  lower_bound(5) = " << *s.lower_bound(5) << endl;
    cout << "  upper_bound(5) = " << *s.upper_bound(5) << endl;

    cout << endl;
}

// === multiset / multimap ===
void multi_demo() {
    cout << "=== multiset / multimap ===" << endl;

    multiset<int> ms = {1, 2, 2, 3, 3, 3};
    cout << "  multiset: ";
    for (int x : ms) cout << x << " ";
    cout << endl;
    cout << "  count(3) = " << ms.count(3) << endl;

    multimap<string, int> mm;
    mm.insert({"a", 1});
    mm.insert({"a", 2});
    mm.insert({"b", 3});

    cout << "  multimap:" << endl;
    for (const auto& [k, v] : mm) {
        cout << "    " << k << " => " << v << endl;
    }

    cout << endl;
}

// === カスタム連結リスト ===
template<typename T>
class LinkedList {
public:
    struct Node {
        T value;
        unique_ptr<Node> next;
        Node(const T& v) : value(v), next(nullptr) {}
    };

    LinkedList() : head_(nullptr), tail_(nullptr), size_(0) {}

    void push_front(const T& value) {
        auto node = make_unique<Node>(value);
        node->next = move(head_);
        head_ = move(node);
        if (!tail_) tail_ = head_.get();
        ++size_;
    }

    void push_back(const T& value) {
        auto node = make_unique<Node>(value);
        auto* raw = node.get();
        if (tail_) {
            tail_->next = move(node);
            tail_ = raw;
        } else {
            head_ = move(node);
            tail_ = raw;
        }
        ++size_;
    }

    void pop_front() {
        if (head_) {
            head_ = move(head_->next);
            if (!head_) tail_ = nullptr;
            --size_;
        }
    }

    T& front() { return head_->value; }
    const T& front() const { return head_->value; }

    size_t size() const { return size_; }
    bool empty() const { return size_ == 0; }

    void print() const {
        cout << "[";
        for (auto* node = head_.get(); node; node = node->next.get()) {
            cout << node->value;
            if (node->next) cout << " -> ";
        }
        cout << "]" << endl;
    }

private:
    unique_ptr<Node> head_;
    Node* tail_;
    size_t size_;
};

void custom_linked_list_demo() {
    cout << "=== カスタム連結リスト ===" << endl;

    LinkedList<int> list;
    list.push_back(1);
    list.push_back(2);
    list.push_back(3);
    list.push_front(0);

    cout << "  ";
    list.print();

    list.pop_front();
    cout << "  after pop_front: ";
    list.print();

    cout << endl;
}

// === 二分探索木 ===
template<typename T>
class BST {
public:
    struct Node {
        T value;
        unique_ptr<Node> left;
        unique_ptr<Node> right;
        Node(const T& v) : value(v), left(nullptr), right(nullptr) {}
    };

    void insert(const T& value) {
        insert_impl(root_, value);
    }

    bool search(const T& value) const {
        return search_impl(root_.get(), value);
    }

    void inorder(vector<T>& result) const {
        inorder_impl(root_.get(), result);
    }

private:
    unique_ptr<Node> root_;

    void insert_impl(unique_ptr<Node>& node, const T& value) {
        if (!node) {
            node = make_unique<Node>(value);
        } else if (value < node->value) {
            insert_impl(node->left, value);
        } else {
            insert_impl(node->right, value);
        }
    }

    bool search_impl(const Node* node, const T& value) const {
        if (!node) return false;
        if (value == node->value) return true;
        if (value < node->value) return search_impl(node->left.get(), value);
        return search_impl(node->right.get(), value);
    }

    void inorder_impl(const Node* node, vector<T>& result) const {
        if (!node) return;
        inorder_impl(node->left.get(), result);
        result.push_back(node->value);
        inorder_impl(node->right.get(), result);
    }
};

void bst_demo() {
    cout << "=== 二分探索木 ===" << endl;

    BST<int> bst;
    for (int v : {5, 3, 7, 1, 4, 6, 8}) {
        bst.insert(v);
    }

    vector<int> sorted;
    bst.inorder(sorted);
    cout << "  inorder: ";
    for (int x : sorted) cout << x << " ";
    cout << endl;

    cout << "  search(4): " << boolalpha << bst.search(4) << endl;
    cout << "  search(9): " << bst.search(9) << endl;

    cout << endl;
}

int main() {
    cout << "=== Data Structures Demo ===" << endl << endl;

    vector_demo();
    array_demo();
    list_demo();
    deque_demo();
    stack_demo();
    queue_demo();
    priority_queue_demo();
    map_demo();
    unordered_map_demo();
    set_demo();
    multi_demo();
    custom_linked_list_demo();
    bst_demo();

    return 0;
}
